File: example-microservice-java/src/main/java/com/example/java/customer/repository/CustomerMapper.java
package com.example.java.customer.repository;

import com.example.java.customer.service.CustomerModel;
import com.example.java.customer.api.dto.CustomerCreateDto;
import com.example.java.customer.api.dto.CustomerUpdateDto;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.mapstruct.factory.Mappers;

@Mapper
public interface CustomerMapper {

    CustomerMapper INSTANCE = Mappers.getMapper(CustomerMapper.class);

    CustomerEntity customerCreateDtoToEntity(CustomerCreateDto source);

    @Mapping(target = "id", ignore = true)
    CustomerEntity customerUpdateDtoToEntity(CustomerUpdateDto source);

    CustomerModel customerEntityToModel(CustomerEntity source);

}

File: example-microservice-kotlin/src/main/kotlin/com/gooddata/kotlin/customer/repository/CustomerMapper.kt
package com.example.kotlin.customer.repository

import com.example.kotlin.customer.api.dto.CustomerCreateDto
import com.example.kotlin.customer.api.dto.CustomerUpdateDto
import com.example.kotlin.customer.service.CustomerModel
import org.mapstruct.Mapper
import org.mapstruct.Mapping
import org.mapstruct.factory.Mappers

@Mapper
interface CustomerMapper {

    companion object {
        val INSTANCE: CustomerMapper = Mappers.getMapper(CustomerMapper::class.java)
    }

    fun customerCreateDtoToEntity(source: CustomerCreateDto): CustomerEntity

    @Mapping(target = "id", ignore = true)
    fun customerUpdateDtoToEntity(source: CustomerUpdateDto): CustomerEntity

    fun customerEntityToModel(source: CustomerEntity): CustomerModel

}
